MCprep_addon/spawner/mcmodel.py

```diff
@@ -363,55 +386,153 @@ def add_model(
+				uvs_base = [p_TL, p_TR, p_BR, p_BL]
+			elif face_name == SOUTH_DIR: # Z+ (V-flip and U-flip)
+				uvs_base = [p_BR, p_BL, p_TL, p_TR]
+			elif face_name == EAST_DIR: # X+ (V-flip and U-flip)
+				uvs_base = [p_BR, p_BL, p_TL, p_TR]
+			elif face_name == WEST_DIR: # X- (V-flip)
+				uvs_base = [p_TL, p_TR, p_BR, p_BL]
+			else:
+				# Default fallback (should not happen, just in case though)
+				uvs_base = [p_TL, p_TR, p_BR, p_BL]
+
+			# Reverse vertex order for downward faces before creation — keeps UVs intact
+			f_for_face = list(f)
+			if face_name == DOWN_DIR:
+				f_for_face = list(reversed(f_for_face))
+
+			if len(f_for_face) != 4:
+				env.log(f"Face vertex index list expected 4 elements, but got {len(f_for_face)}. Skipping face.")
+				continue
+
+			try:
+				face_verts = []
+				for vert_idx in f_for_face:
+					if vert_idx < 0 or vert_idx >= len(verts):
+						env.log(f"Vertex index {vert_idx} is out of bounds (0 to {len(verts)-1}). Skipping face.")
+						raise IndexError("Vertex index out of bounds")
+					face_verts.append(verts[vert_idx])
+
+				face = bm.faces.new(tuple(face_verts))
+			except ValueError as e:
+				# If face cannot be created, skip it
+				env.log(f"Failed to create BMesh face: {e}. Skipping.")
+				continue
+			except IndexError:
+				# Catch the custom index error from the bounds check
+				continue
 
 			face.normal_update()
 
 			# Give slight offset by normal for overlay geometry
 			if face_mat == "#overlay":
-				bmesh.ops.translate(bm, verts=face.verts,
-									vec=0.0025 * face.normal)
-
-			for j in range(len(face.loops)):
-				# uv coords order is determened by the rotation of the uv,
-				# e.g. if the uv is rotated by 180 degrees, the first index
-				# will be 2 then 3, 0, 1.
-				face.loops[j][uv_layer].uv = uvs[(j + uv_idx) % len(uvs)]
+				bmesh.ops.translate(bm, verts=face.verts, vec=0.0025 * face.normal)
+
+			# Apply rotation shift (uv_idx)
+			uvs_rot = [uvs_base[(k + uv_idx) % len(uvs_base)] for k in range(len(uvs_base))]
+
+			# Mirror axis per-face (to fix convention mismatches)
+			mirror_axis = None
+			if face_name == NORTH_DIR:
+				mirror_axis = "Y"
+			elif face_name == EAST_DIR:
+				mirror_axis = "X"
+			elif face_name == SOUTH_DIR:
+				mirror_axis = "X"
+			elif face_name == WEST_DIR:
+				mirror_axis = "Y"
+			elif face_name == DOWN_DIR:
+				mirror_axis = "Y"
+			# up -> no mirroring
+
+			# If mirroring, get UV island median and mirror around point.
+			if mirror_axis is not None:
+				mx = sum(p[0] for p in uvs_rot) / len(uvs_rot)
+				my = sum(p[1] for p in uvs_rot) / len(uvs_rot)
+				uvs_final = []
+				for (x, y) in uvs_rot:
+					if mirror_axis == "X":
+						nx = 2 * mx - x
+						ny = y
+					else:
+						nx = x
+						ny = 2 * my - y
+					uvs_final.append((nx, ny))
+			else:
+				uvs_final = uvs_rot
+
+			# For cardinal faces (north/south/east/west) mirror over both axes
+			if face_name in {NORTH_DIR, SOUTH_DIR, EAST_DIR, WEST_DIR}:
```

Comment:
This can be done as a tuple instead of a dictionary


```suggestion
			if face_name in (NORTH_DIR, SOUTH_DIR, EAST_DIR, WEST_DIR):
```